File: backend/python/email_processor/utils/email_classifier.py
import re
from config.db_config import get_db_connection


def classify_email(email):
    """
    Classifies an email based on whitelist/blocklist and classification rules.

    Args:
        email (dict): The email to classify.

    Returns:
        str: Classification status.
    """
    connection = get_db_connection()
    cursor = connection.cursor(dictionary=True)

    try:
        # Fetch whitelist and blocklist entries
        cursor.execute("SELECT * FROM emaillistentries WHERE status = 1")
        email_list_entries = cursor.fetchall()

        # Check blocklist
        for entry in email_list_entries:
            if entry["type"] == "blocklist":
                if entry["email_address"] and entry["email_address"].lower() == email["from"].lower():
                    return "Filtered Out"
                if entry["domain"] and email["from"].split("@")[-1].lower() == entry["domain"].lower():
                    return "Filtered Out"

        # Check whitelist
        for entry in email_list_entries:
            if entry["type"] == "whitelist":
                if entry["email_address"] and entry["email_address"].lower() == email["from"].lower():
                    return "Confirmed Order Email"
                if entry["domain"] and email["from"].split("@")[-1].lower() == entry["domain"].lower():
                    return "Confirmed Order Email"

        # Fetch classification rules
        cursor.execute("SELECT * FROM classificationrules WHERE is_active = 1")
        classification_rules = cursor.fetchall()

        # Apply classification rules
        for rule in classification_rules:
            # Subject-based rules
            if rule["type"] == "Subject" and re.search(rule["pattern"], email["subject"], re.IGNORECASE):
                return "Confirmed Order Email"

            # Body text-based rules
            if rule["type"] == "Body Text" and re.search(rule["pattern"], email["text"], re.IGNORECASE):
                return "Confirmed Order Email"

            # Attachment name-based rules
            if rule["type"] == "Attachment Name":
                for attachment in email.get("attachments", []):
                    if re.search(rule["pattern"], attachment["filename"], re.IGNORECASE):
                        return "Confirmed Order Email"

        # Default classification
        return "Pending Review"

    except Exception as e:
        print(f"Error in classification: {e}")
        return "Pending Review"
    finally:
        cursor.close()
        connection.close()

File: backend/utils/emailClassifier.js
const { ClassificationRule, EmailAttachment, EmailListEntry } = require('../models');

const classifyEmail = async (email) => {
  // Check against email filtering rules (whitelist/blocklist)
  const isWhitelisted = await EmailListEntry.findOne({
    where: { type: 'whitelist', email_address: email.sender, status: true },
  });
  const isBlacklisted = await EmailListEntry.findOne({
    where: { type: 'blocklist', email_address: email.sender, status: true },
  });

  if (isBlacklisted) {
    return { status: 'Filtered Out', reason: 'Blacklisted email sender' };
  }

  if (isWhitelisted) {
    return { status: 'Confirmed Order Email', reason: 'Whitelisted email sender' };
  }

  // Check against classification rules
  const rules = await ClassificationRule.findAll({ where: { is_active: true } });

  for (const rule of rules) {
    const pattern = new RegExp(rule.pattern, 'i');
    if (rule.type === 'Subject' && pattern.test(email.subject)) {
      return { status: 'Confirmed Order Email', reason: `Matched subject rule: ${rule.rule_name}` };
    } else if (rule.type === 'Body Text' && pattern.test(email.content)) {
      return { status: 'Confirmed Order Email', reason: `Matched body text rule: ${rule.rule_name}` };
    } else if (rule.type === 'Attachment Name') {
      const attachments = await EmailAttachment.findAll({ where: { email_id: email.id } });
      const attachmentMatches = attachments.some((attachment) => pattern.test(attachment.file_name));
      if (attachmentMatches) {
        return { status: 'Confirmed Order Email', reason: `Matched attachment rule: ${rule.rule_name}` };
      }
    }
  }

  // Default to pending review if no rules matched
  return { status: 'Pending Review', reason: 'No matching rules found' };
};

module.exports = classifyEmail;
